src/Tools/dotnet-monitor/StartupHook/StartupHookValidator.cs

```diff
@@ -34,43 +33,88 @@ internal sealed class StartupHookValidator
             _sharedLibraryService = sharedLibraryService;
         }
 
-        public async Task<bool> CheckAsync(IEndpointInfo endpointInfo, CancellationToken token)
+        public async Task<bool> ApplyStartupHook(IEndpointInfo endpointInfo, CancellationToken token)
         {
-            IFileProviderFactory fileProviderFactory = await _sharedLibraryService.GetFactoryAsync(token);
+            bool startupHookEnvVarSet = await CheckStartupHookEnvVarAsync(endpointInfo, token);
 
-            IFileProvider managedFileProvider = fileProviderFactory.CreateManaged(StartupHookTargetFramework);
+            if (startupHookEnvVarSet)
+            {
+                return true;
+            }
+
+            if (endpointInfo.RuntimeVersion.Major >= 8)
+            {
+                try
+                {
+                    IFileInfo startupHookLibraryFileInfo = await GetStartupHookLibaryFileInfo(token);
+                    DiagnosticsClient client = new(endpointInfo.Endpoint);
+
+                    await client.ApplyStartupHookAsync(startupHookLibraryFileInfo.PhysicalPath, token);
+
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    _logger.StartupHookApplyFailed(ex);
+                }
+            }
+
+            return false;
+        }
 
-            IFileInfo startupHookLibraryFileInfo = managedFileProvider.GetFileInfo(StartupHookFileName);
-            if (!startupHookLibraryFileInfo.Exists)
+        public async Task<bool> CheckAsync(IEndpointInfo endpointInfo, CancellationToken token, bool logInstructions = false)
+        {
+            if (endpointInfo.RuntimeVersion.Major >= 8)
             {
-                // This would be a bug in dotnet-monitor; throw appropriate non-MonitoringException instance.
-                throw new FileNotFoundException(null, startupHookLibraryFileInfo.Name);
+                // We currently have no way of tracking whether the startup hook was applied successfully; for now,
```

Comment:
Joe and I chatted about this a bit yesterday - we currently have no mechanism for tracking whether `ApplyStartupHookAsync` succeeded when we call `CheckAsync`, so for now this operates on the assumption that it did succeed. I believe the consensus was this is something that probably should be addressed, but that it's outside the scope of this change.

I can see a bit of an interesting edge-case here though - if someone has the env var set on 8+, we respect that and don't attempt to call `ApplyStartupHookAsync`. However, if that fails, it won't be reflected here, since we just assume 8 succeeds. We could probably add some more nuanced logic to this check that looks for the env var - if a value is there, we check to ensure it appears valid; if not, we blindly return true.